pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/IndexService.java

```diff
@@ -0,0 +1,83 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.pinot.segment.spi.index;
+
+import com.google.common.collect.Sets;
+import java.util.HashSet;
+import java.util.Optional;
+import java.util.ServiceLoader;
+import java.util.Set;
+
+
+/**
+ * This is the entry point of the Index SPI.
+ *
+ * Ideally, if we used some kind of injection system, this class should be injected into a Pinot context all classes can
+ * receive when they are built. Given that Pinot doesn't have that, we have to relay on static fields.
```

Comment:
These comments will be useful when added as a doc for `_instance` member variable.

Reply:
TL;DR: I think it is better to keep it here, but I don't have an strong opinion about that and we can move it if consensus.

Why I think it is better to be here:

I don't like the idea of adding these comments into a private attribute. Sure, it may be slightly better for someone that reads the class, but these lines may also be useful for callers. I would expect callers ask why they have to use this class by doing `IndexService.getInstance().getAllIndexes()`, which is quite verbose, instead of simply calling `IndexService.getAllIndexes()`. Well, they can answer that question is that we hope that, in a (probably far) future, you could simply inject the IndexService instance in your constructor. A caller can understand that if they open the Javadoc of `IndexService` in their own IDE without actually needed to read the source code of `IndexService`. 

If we move these lines to the attribute javadoc instead of the class javadoc, the caller would need to open the class, hope that they have the sources downloaded and then go to some private attribute they don't actually know their existence in order to read that.

At the end of the day this comment mixes information related to the usability (why we decided to show this interface) and the implementation (why we didn't use other ways). The former is interesting for callers and therefore should be attached to some public concept (like the class or the getInstance method) and the later is interesting for the implementator (and therefore may be attached to the private attribute).

Reply:
I think it is well-known to a Pinot developer that we don't use a standard injection mechanism across the codebase.  My comment was specifically for these 2 lines:

```Ideally, if we used some kind of injection system, this class should be injected into a Pinot context all classes can receive when they are built. Given that Pinot doesn't have that, we have to relay on static fields.``` 

It is not clear what static field you are talking about. When the interface evolves in the future, this comment might no longer be relevant. 

Imo, it seems out of place as it doesn't talk about the usability of the interface. It pertains to the current choice of implementation that is not required for any caller or even implementer to know because it is well-known to a developer that Pinot doesn't use a standard injection mechanism across the codebase. 

If you want to emphasize the usability of this interface, you should just literally explain how to use it in the javadoc. The future improvements can be annotated as a TODO or Note. 

An example could be:

``` 
This is the entry point of the Index SPI.
 
 By default, this service will be initialized by reading all ServiceLoader SPI services 
that implement {@link IndexPlugin}, adding all the {@link IndexType} that can be found in 
that way. A caller can iterate it by invoking `IndexService.getInstance().getAllIndexes()` 
 
The interface allows changing the singleton instance at runtime using the method {@link #setInstance(IndexService)}. 

Note:  Since Pinot doesn't support injection mechanism, we have to rely on static fields (singleton pattern) like `_instance`. With an injection mechanism, this instance can be part of a `PinotContext` object that is accessible in all components.

```

I still think those 2 lines in the javadoc are out of place and me, as a pinot developer, don't find it useful. Will let others chime in.

Reply:
I've made some changes to the javadoc. I hope it is better now! :)